shared-bindings/alarm/__init__.c

```diff
@@ -0,0 +1,239 @@
+#include "py/reload.h"
+#include "py/runtime.h"
+
+#include "shared-bindings/alarm/__init__.h"
+#include "shared-bindings/alarm/pin/PinAlarm.h"
+#include "shared-bindings/alarm/time/TimeAlarm.h"
+#include "shared-bindings/supervisor/Runtime.h"
+#include "shared-bindings/time/__init__.h"
+#include "supervisor/shared/autoreload.h"
+#include "supervisor/shared/workflow.h"
+
+// Wait this long imediately after startup to see if we are connected to USB.
+#define CIRCUITPY_USB_CONNECTED_SLEEP_DELAY 5
+
+//| """Alarms and sleep
+//|
+//| Provides alarms that trigger based on time intervals or on external events, such as pin
+//| changes.
+//| The program can simply wait for these alarms, or go into a sleep state and
+//| and be awoken when they trigger.
+//|
+//| There are two supported levels of sleep: light sleep and deep sleep.
+//|
+//| Light sleep leaves the CPU and RAM powered so the program can resume after sleeping.
+//|
+//| *However, note that on some platforms, light sleep will shut down some communications, including
+//| WiFi and/or Bluetooth.*
+//|
+//| Deep sleep shuts down power to nearly all of the microcontroller including the CPU and RAM. This can save
+//| a more significant amount of power, but CircuitPython must restart ``code.py`` from the beginning when
+//| awakened.
+//| """
+
+//|
+//| wake_alarm: Alarm
+//| """The most recently triggered alarm. If CircuitPython was sleeping, the alarm the woke it from sleep."""
+//|
+
+// wake_alarm is implemented as a dictionary entry, so there's no code here.
+
+void validate_objs_are_alarms(size_t n_args, const mp_obj_t *objs) {
+    for (size_t i = 0; i < n_args; i++) {
+        if (MP_OBJ_IS_TYPE(objs[i], &alarm_pin_pin_alarm_type) ||
+            MP_OBJ_IS_TYPE(objs[i], &alarm_time_time_alarm_type)) {
+            continue;
+        }
+        mp_raise_TypeError_varg(translate("Expected an alarm"));
+    }
+}
+
+//| def wait_until_alarms(*alarms: Alarm) -> Alarm:
+//|     """Wait for one of the alarms to trigger. The triggering alarm is returned.
+//|     is returned, and is also available as `alarm.wake_alarm`. Nothing is shut down
+//|     or interrupted. Power consumption will be reduced if possible.
+//|
+//|     If no alarms are specified, return immediately.
+//|     """
+//|     ...
+//|
+STATIC mp_obj_t alarm_wait_until_alarms(size_t n_args, const mp_obj_t *args) {
+    validate_objs_are_alarms(n_args, args);
+    common_hal_alarm_wait_until_alarms(n_args, args);
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(alarm_wait_until_alarms_obj, 1, MP_OBJ_FUN_ARGS_MAX, alarm_wait_until_alarms);
+
+//| def sleep_until_alarms(*alarms: Alarm) -> Alarm:
+//|     """Go into a light sleep until awakened one of the alarms. The alarm causing the wake-up
+//|     is returned, and is also available as `alarm.wake_alarm`.
+//|
+//|     Some functionality may be shut down during sleep. On ESP32-S2, WiFi is turned off,
+//|     and existing connections are broken.
```

Comment:
I'd prefer not to do this. Instead, we should just have one light sleep that works like `wait_until_alarms`.

This statement leaves too much uncertainty about what needs to happen after this function to get things going again.

Reply:
(discussion in lines 53-54 comment)